renderapi/client.py

```diff
@@ -19,6 +20,61 @@
 logger.addHandler(NullHandler())
 
 
+@decorator
+def renderclientaccess(f, *args, **kwargs):
+    """Decorator allowing functions asking for host, port, owner, project,
+    client_script to default to a connection defined by :class:`RenderClient`
+    object using its :func:`RenderClient.make_kwargs` method.
+    Will also attempt to derive a :class:`RenderClient` from an input
+    :class:`Render` object and fail if client scripts cannot be reached.
+
+    Parameters
+    ----------
+    f : func
+        function to decorate
+    Returns
+    -------
+    obj
+        output of decorated function
+    """
+    args, kwargs = fitargspec(f, args, kwargs)
+    render = kwargs.get('render')
+    if render is not None:
+        if not isinstance(render, RenderClient):
+            if isinstance(render, Render):
+                render = RenderClient(**render.make_kwargs(**kwargs))
+            else:
+                raise ValueError(
+                    'invalid RenderClient object type {} specified!'.format(
+                        type(render)))
+        return f(*args, **render.make_kwargs(**kwargs))
+    else:
+        try:
+            client_script = kwargs.get('client_script')
+            cs_valid = os.path.isfile(client_script)
+        except TypeError as e:
+            try:
+                if os.path.isdir(kwargs.get('client_scripts')):
+                    client_script = os.path.join(client_scripts,
+                                                 'run_ws_client.sh')
+                    cs_valid = os.path.isfile(client_script)
+                else:
+                    raise ClientScriptError(
+                        'invalid client_scripts directory {}'.format(
+                            kwargs.get('client_scripts')))
+            except TypeError as e:
+                raise ClientScriptError(
+                    'No client script information specified: '
+                    'client_scripts={} client_script={}'.format(
+                        kwargs.get('client_scripts'),
+                        kwargs.get('client_script')))
+        if not cs_valid:
+            # TODO should also check for executability
+            raise ClientScriptError(
+                'invalid client script: {} not a file'.format(client_script))
+    return f(*args, **kwargs)
+

```

Comment:
An error we have been getting is when the client_scripts are specified, but that script is not in the cannonical part of a render repository that has setup a 'deploy' directory.  When this happens, the script exists, but the script when run will either assume that the jdk and the jar can be found in the canonical location in the repo, or will check for the environment variables RENDER_JAVA_HOME and RENDER_CLIENT_JAR to tell it where to find the jar and the java_home.  So I think that we should probably put that validation in here as well... i.e. check that the jdk and jar exist in the cannonical location or the environment variables exist so this runs correctly.  This was in fact the problem that Tim had run into a couple times where he forgot to specify these environment variables and then the solver was passed that script and it didn't really error appropriately.  Our schemas already validate that the script is a file that exists.

Reply:
The schemas only validate at the client_scripts directory level, and renderapi.connect does not fail in the case of a missing client script by default.  This solves a different issue that Gayathri and Tim have had previously.

I was avoiding making assumptions about the content of the client script (and this commit made it more general rather than hardcoded) so that it could be more flexible, for instance, if one wanted to try to work with a Windows version of these tools -- I thought I saw some work being done on that.

It is fine to add in this validation, but I'm curious if it wouldn't make more sense to exit with a more informative message in setup_java_env.sh

Reply:
yeah .. maybe the right thing to do is have setup_java_env.sh have a set -e at the top so that is quits with an error.  I think the problem i had was whether the matlab solver called the script and it exited, i'm not sure with or without error, but there was no message coming back that the script couldn't find the jar or the jdk because it wasn't in the canonical location.

Reply:
actually.. similar but different just the setup_java_env should validate that the RENDER_JAVA_HOME and RENDER_CLIENT_JAR that are set when its done are valid java_home and RENDER_CLIENT_JAR files .. that they exist at least.  I'll suggest this in render instead.